### DISCUSSION TITLE
AlphaTab.Model.JsonConverter available in C#?

### DISCUSSION THREAD

**Discussion by rhclayto** [CoderLine/alphaTab#425] [Q&A] [answered]
# Question 
Is `AlphaTab.Model.JsonConverter` available in the C# version available from Nuget? I see in an old issue that it seems to be ( https://github.com/CoderLine/alphaTab/issues/87#issuecomment-208006767 ), but when I try to use it in a C# project in Visual Studio, it's not available.

```csharp
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using AlphaTab.Model;

namespace ConsoleApp1
{
    class Program
    {
        static void Main(string[] args)
        {
            // var score = ... // Omitted for brevity.
            // This gives the error in VS: Error CS0234 The type or namespace name 'JsonConverter'
            // does not exist in the namespace 'AlphaTab.Model' (are you missing an assembly reference?)        
           var json = AlphaTab.Model.JsonConverter.ScoreToJson(score);
        }
    }
}
```

# Your environment (if applicable)
* Version used: Nuget package
* Platform used: C#
* Operating System and version (desktop or mobile): Windows 10, desktop, Visual Studio 2019

**Comment by Danielku15** [comment]
The JsonConverter is a component specific to the web target. it is mainly used to transport various objects from the main browser thread to the web worker in the back. So far there was no need to bring the JsonConverter to the .net platform as it is mostly meant for the mentioned usecase. 

What would be the actual usecase for you to have JSON serialization? The most convenient way of "writing" new piece with alphaTab is definitely alphaTex. The JSON model might work, but is quite sensitive to mistakes. If your application provides writing new music pieces (by manipulating the data model directly) there is no built-in exporter so far. I once maintained an alphaTex exporter beside the importer but it was not used so I dropped it due to maintenance overhead. 

The "model editing APIs" and "file exporters" topics are rather deep down at the end of the backlog for [alphaTab 1.5
](https://github.com/orgs/CoderLine/projects). The problem with the JSON serializer is that it is not versionized or backwards compatible enough to be fully suitable for storing generated scores. 

To answer the question of this issue: The JsonConverter is and will likely not become available on .net. If you have some concrete needs (e.g. to save scores) feel free to open a feature request dedicated to that.

**Comment by rhclayto** [comment]
Thanks for your reply.

My use case is similar to the use case you described in that it involves being able to pass score objects between different components:
1) Storing scores as JSON in a database server-side, then serving them as JSON to browser clients from a REST API. JavaScript is fine for this because I can use node.js on the server.
2) Sharing the same score between AlphaTab C# & AlphaTab JavaScript. The client is a Unity game, that also includes an embedded browser which handles score rendering & synth playing. (I found it too hard to use the AlphaTab low-level APIs with Unity's GUI system.) The score needs to be available in both the Unity C# environment & the browser JavaScript environment. I thought a solution would be to pass JSON between them, as in the browser with WebWorkers. Right now the Unity side is acting as the REST API client, but if I change that to use the browser to fetch the JSON from the server I can pass that to Unity. So I suppose I don't need a C# JsonConverter after all.

Let me ask about AlphaTex though. Is it capable of representing all the same information as a Guitar Pro file? For instance, if you took a .gp file, got the Score model from it with AlphaTab, then--assuming there was an AlphaTex exporter--could you export AlphaTex without losing any information?

**Comment by Danielku15** [comment]
1. The need to store user created files is definitly a valid usecase. But for storing it in the DB the JSON schema of alphaTab is not reliable enough. For transmitting a score from a backend to the frontend would work as long as frontend+backend share the same alphaTab version but beyond that I would not rely on its backwards+forwards compatibility. 
2. There were some people requesting in the past some Unity support. When I looked at it in the past I could not really find a proper GUI library for Unity that could be targeted. Hence the low level APIs where you get some rendered bitmaps or SVGs is usually the best choice for such platforms. I'm happy to receive feature requests for some more concrete libraries there. 

> Let me ask about AlphaTex though. Is it capable of representing all the same information as a Guitar Pro file? 

I would say mostly yes. There might be some missing small bits here and there but most features alphaTab supports, also alphaTex provides as input syntax. Any missing piece should be added quickly in case somebody would need it 😉
I guess only creating an exporter would really tell me if there might be some data in the data model which does not have an alphaTex comparable syntax. 

For alphaTab 1.1 I plan to not add any big features anymore, but the scope of 1.2 is not really fixed. I could also refocus and create an exporter in 1.2 (or an 1.1.1). In the end it depends on what brings the most value to the integrators. 

- alphaTex export would be mostly for alphaTab only uses. I guess there might not be so many usecases this is beneficial. 
- GuitarPro7 export would be nice and there are already quite some tools which support the format. 
- MusicXML I am not really a fan of, the format is too generic and every application does a bit different. 
- Music Encoding Initiative (MEI) would be quite nice and the importer is already on my roadmap. But it is not yet widely adopted so other applications might not yet benefit much from it. I also did not check yet in detail if really all elements would be supported. 

 So I guess a GuitarPro7 export would bring the biggest value in short term.

**Comment by Danielku15** [comment]
FYI: I added #443 as work item for 7.2 which should allow both usecases: storing dynamically generated songs and transmitting files over the wire. I will check if the JSON serializer would be sort-of compatible with C# as part of this issue. Unfortunately this will likely pull a new dependency `Newtonsoft.Json` or `System.Text.Json` into the .net version of alphaTab.

**Comment by Danielku15** [comment, marked as answer]
Just transferred this question to the new GitHub Discussions and I would like to close it with the following answer:

In version 1.1 there is no good way to save files or exchange it between different platforms. I entered following items for 1.2:

- For the file saving scenario #443 will bring a Guitar Pro 7 exporter. 
- For the data exchange scenario #461 will bring the JSON serialization to .net

**Reply by Danielku15** [reply]
@rhclayto The `JsonConverter` is now also available on C#. The JsonConverter will just generate a `Map<string, object>` (IDictionary) which can then be passed on to another serializer like JSON.net. Give it a try in 1.2.0-alpha.74
https://www.alphatab.net/docs/guides/lowlevel-apis/#serialize-data-model-fromto-json

**Reply by Danielku15** [reply]
And now the GP7 Exporter is also available in 1.2.0-alpha.75
Related docs https://www.alphatab.net/docs/guides/exporter
